kr_mesh_visualization/src/mesh_visualization.cpp

```diff
@@ -1,83 +1,100 @@
-#include <geometry_msgs/PoseStamped.h>
-#include <geometry_msgs/PoseWithCovarianceStamped.h>
-#include <nav_msgs/Odometry.h>
-#include <ros/ros.h>
-#include <topic_tools/shape_shifter.h>
-#include <visualization_msgs/Marker.h>
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
+#include "nav_msgs/msg/odometry.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "visualization_msgs/msg/marker.hpp"
 
-static std::string mesh_resource, new_frame_id;
-static ros::Publisher pub_vis;
-static double color_r, color_g, color_b, color_a;
-static double scale_x, scale_y, scale_z;
-
-static void publishMarker(const std::string &frame_id, const geometry_msgs::Pose &pose)
+class MeshVisualizationNode : public rclcpp::Node
 {
-  visualization_msgs::Marker marker;
-  marker.header.frame_id = (new_frame_id == "") ? frame_id : new_frame_id;
-  marker.header.stamp = ros::Time();  // time 0 so that the marker will be
-                                      // displayed regardless of the current time
-  marker.ns = ros::this_node::getName();
-  marker.id = 0;
-  marker.type = visualization_msgs::Marker::MESH_RESOURCE;
-  marker.action = visualization_msgs::Marker::ADD;
-  marker.pose = pose;
-  marker.scale.x = scale_x;
-  marker.scale.y = scale_y;
-  marker.scale.z = scale_z;
-  marker.color.a = color_a;
-  marker.color.r = color_r;
-  marker.color.g = color_g;
-  marker.color.b = color_b;
-  marker.mesh_resource = mesh_resource;
-  pub_vis.publish(marker);
-}
+ public:
+  MeshVisualizationNode() : Node("kr_mesh_visualization")
+  {
+    this->declare_parameter<std::string>("mesh_resource",
+                                         std::string("package://kr_mesh_visualization/mesh/hummingbird.mesh"));
+    this->declare_parameter<double>("color/r", 1.0);
+    this->declare_parameter<double>("color/g", 0.0);
+    this->declare_parameter<double>("color/b", 0.0);
+    this->declare_parameter<double>("color/a", 1.0);
+    this->declare_parameter<double>("scale/x", 1.0);
+    this->declare_parameter<double>("scale/y", 1.0);
+    this->declare_parameter<double>("scale/z", 1.0);
+    this->declare_parameter<std::string>("new_frame_id", std::string(""));
 
-static void any_callback(const topic_tools::ShapeShifter::ConstPtr &msg)
-{
-  // ROS_INFO_STREAM("Type: " << msg->getDataType());
+    this->get_parameter("mesh_resource", mesh_resource);
+    this->get_parameter("color/r", color_r);
+    this->get_parameter("color/g", color_g);
+    this->get_parameter("color/b", color_b);
+    this->get_parameter("color/a", color_a);
+    this->get_parameter("scale/x", scale_x);
+    this->get_parameter("scale/y", scale_y);
+    this->get_parameter("scale/z", scale_z);
+    this->get_parameter("new_frame_id", new_frame_id);
 
-  if(msg->getDataType() == "geometry_msgs/PoseStamped")
+    using namespace std::placeholders;
```

Comment:
Is there any reason for using a standard namespace?

Reply:
Yes, it is just so that while making the subscriber the function line doesn't become too long.
Without the namespace the below subscriber becomes:
    sub_odom_ = this->create_subscription<nav_msgs::msg::Odometry>("~/input", 10,
                std::bind(&MeshVisualizationNode::odom_callback, this, std::placeholders::_1));

Reply:
Usually using std namespace is a bad practice https://dev.to/iamcymentho/using-namespace-std-in-c-why-its-considered-bad-practice-2adf

Reply:
Yes, agreed. which is why I have specifically put using using namespace std::placeholders and not the standard std. For example the code [here](https://docs.ros.org/en/humble/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.html) using it in similar fashion. I can remove it if required.